File: binv10/seventh_problem.js
import fs from "fs";

const staffs = JSON.parse(fs.readFileSync("input_data/staff.json", "utf8"));

const latestFeedbacks = staffs.map((staff) => {
  const feedbacks = staff?.result?.feedback;

  const latestFeedback = feedbacks?.reduce((prev, curr) => {
    return new Date(prev.date) > new Date(curr.date) ? prev : curr;
  }, feedbacks[0]);

  return {
    id: staff.id,
    full_name: staff.full_name,
    email: staff.email,
    latest_feedbacks: latestFeedback?.comment ?? "No available feedback yet.",
  };
});

console.log(JSON.stringify({ data: latestFeedbacks }, null, 2));

File: binsV1/binV5/seventh_problem.rb
require 'json'
require 'date'

staffs = JSON.parse(File.read('input_data/staff.json'), symbolize_names: true)

results = staffs.map do |staff|
  feedbacks = staff.dig(:result, :feedback).max_by do |feedback|
    Date.parse(feedback[:date])
  end

  {
    email: staff[:email],
    full_name: staff[:full_name],
    latest_feedback: feedbacks&.fetch(:comment, 'No available feedback yet.')
  }
end

print JSON.pretty_generate({ data: results })
